### DISCUSSION TITLE
2player selection screen

### DISCUSSION THREAD

**Discussion by Raptorq1** [Techokami/SonicWorldsNext#212] [Q&A]
How would I make player 1 and player 2 select separate players

**Comment by r6915ee** [comment]
I mean, it depends on how the scene is created. You could try making separate "modes" of the character select screen to see what each player picks; most games have that functionality.

**Reply by Raptorq1** [reply]
Okay I get it but adding the functionality for each player to select a player that's the part I don't under stand oh and I tried to do it on my own but it didn't work

**Reply by r6915ee** [reply]
You could try setting up additional input map controls and use them instead if it has been detected that there are multiple devices being used.

**Reply by Raptorq1** [reply]
Okay but how would I make it check if both players have selected a character it switches the scene

**Reply by r6915ee** [reply]
If you've done it separately, you can just make it switch the scene afterwards. Otherwise, you can add boolean values that define whether or not a player is "ready" and check if both are ready, then switch the scene, either automatically or on input.

**Reply by Raptorq1** [reply]
Okay but now how would I make the level show up on both veiwports

**Reply by r6915ee** [reply]
As in the level icon, or the actual level itself?

**Reply by Raptorq1** [reply]
The actual level oh and I have written the character select code for both player 1 and player 2 but player 2 isn't changing to the selected character and player 1 isn't working as expected

Player 1 code


extends Node2D

var characterLabels = ["Sonic", "Tails", "Knuckles", "Amy","Mighty","Ray"]
var selected = false
var characterID =0
var lastInput = Vector2.ZERO





func _input(event):
	if !selected:
		
		var inputCue = Input.get_vector("gm_left","gm_right","gm_up","gm_down")
		inputCue.x = round(inputCue.x)
		inputCue.y = round(inputCue.y)
		if inputCue != lastInput:
			if inputCue.x < 0:
					characterID = wrapi(characterID-1,0,characterLabels.size())
					$"../../Switch".play()
			elif inputCue.x > 0:
					characterID = wrapi(characterID+1,0,characterLabels.size())
					$"../../Switch".play()
		lastInput = inputCue
		
		$PlayerName1.text = characterLabels[characterID]
		
		if event.is_action_pressed("gm_action"):
			selected = true
			
			
			
		match(characterID):
				1: # Sonic
					Global.PlayerChar1 = Global.CHARACTERS.SONIC
				2: # Tails
					Global.PlayerChar1 = Global.CHARACTERS.TAILS
				3: # Knuckles
					Global.PlayerChar1 = Global.CHARACTERS.KNUCKLES
				4: # Amy
					Global.PlayerChar1 = Global.CHARACTERS.AMY

Player 2 code


extends Node2D

var characterLabels = ["Sonic", "Tails", "Knuckles", "Amy","Mighty","Ray"]
var selected = false
var characterID =0
var lastInput = Vector2.ZERO

func _input(event):
	if !selected:
		
		var inputCue = Input.get_vector("gm_left_P2","gm_right_P2","gm_up_P2","gm_down_P2")
		inputCue.x = round(inputCue.x)
		inputCue.y = round(inputCue.y)
		if inputCue != lastInput:
			if inputCue.x < 0:
				characterID = wrapi(characterID-1,0,characterLabels.size())
				$"../../Switch".play()
			elif inputCue.x > 0:
				characterID = wrapi(characterID+1,0,characterLabels.size())
				$"../../Switch".play()
		lastInput = inputCue
		
		$PlayerName2.text = characterLabels[characterID]
		
		# finish character select if start is pressed
		if event.is_action_pressed("gm_action_P2"):
			selected = true
			
			
			
		match(characterID):
				1: 
					Global.PlayerChar2 = Global.CHARACTERS.SONIC
				2: 
					Global.PlayerChar2 = Global.CHARACTERS.TAILS
				3: 
					Global.PlayerChar2 = Global.CHARACTERS.KNUCKLES
				4: 
					Global.PlayerChar2 = Global.CHARACTERS.AMY

**Reply by r6915ee** [reply]
First off, since the scripts are very similar to each other, you may want to combine the two into one script and define an exported variable stating what player number it is and use that accordingly.

As for what is going on in the former from your question? Well, there does seem to be extra indentation, particularly regarding `inputCue`. GDScript dislikes extra indentations; the editor should give you an error whenever such occurs.

I haven't tested this, but this would probably be a good way to merge the two scripts and fix the indentation. I'm not sure about what you mentioned with player 2, however:

```gdscript
extends Node2D
# I know that the original two scripts did not have a class name, but I included one here for clarity
class_name PlayerSelectChar

# The character names, in order
var characterLabels := ["Sonic", "Tails", "Knuckles", "Amy", "Mighty", "Ray"]
# Was the player character selected?
var selected := false
# Used for identifying who should be the player character
var characterID := 0
# Self-explanatory
var lastInput := Vector2.ZERO

# Defines what player number should be used for identification
@export var player_num := 0

func _input(event : InputEvent):
	if !selected:
		var inputCue

		if player_num == 0: inputCue = Input.get_vector("gm_left", "gm_right", "gm_up", "gm_down")
		else: inputCue = Input.get_vector("gm_left_P2", "gm_right_P2", "gm_up_P2", "gm_down_P2")
		inputCue.x = round(inputCue.x)
		inputCue.y = round(inputCue.y)

		if inputCue != lastInput:
			if inputCue.x < 0:
				characterID = wrapi(characterID-1,0,characterLabels.size())
				$"../../Switch".play()
			elif inputCue.x > 0:
				characterID = wrapi(characterID+1,0,characterLabels.size())
				$"../../Switch".play()
		lastInput = inputCue

		if player_num == 0: $PlayerName1.text = characterLabels[characterID]
		else: $PlayerName2.text = characterLabels[characterID]
		
		# Finish character select if start is pressed
		# When a body of code (made by statements with : such as an if statement) is only
		# one line of code, it can be put on the same line as the statement being used.
		# The same thing happens here. It's also assumed that Mighty and Ray are already present
		# in Global.CHARACTERS.
		if player_num == 0:
			if event.is_action_pressed("gm_action"):
				selected = true
			
				match(characterID):
					1: Global.PlayerChar1 = Global.CHARACTERS.SONIC
					2: Global.PlayerChar1 = Global.CHARACTERS.TAILS
					3: Global.PlayerChar1 = Global.CHARACTERS.KNUCKLES
					4: Global.PlayerChar1 = Global.CHARACTERS.AMY
					5: Global.PlayerChar1 = Global.CHARACTERS.MIGHTY
					6: Global.PlayerChar1 = Global.CHARACTERS.RAY
		else:
			if event.is_action_pressed("gm_action_P2"):
				selected = true
			
				match(characterID):
					1: Global.PlayerChar2 = Global.CHARACTERS.SONIC
					2: Global.PlayerChar2 = Global.CHARACTERS.TAILS
					3: Global.PlayerChar2 = Global.CHARACTERS.KNUCKLES
					4: Global.PlayerChar2 = Global.CHARACTERS.AMY
					5: Global.PlayerChar2 = Global.CHARACTERS.MIGHTY
					6: Global.PlayerChar2 = Global.CHARACTERS.RAY
```

As for the level, is it supposed to be in split-screen, or via a different method?

**Reply by Raptorq1** [reply]
I have a question to what node would I connect the script to oh and how would I make it so that player 2 doesn't follow player 1 and make the HUD get displayed for both players and player 1 and 2 each have there own score counter and time counter and ring counter

**Reply by r6915ee** [reply]
It generally depends on how you structure your scene. If the nodes do not exist yet, you can add them like so **in code** in the parent node's script:

```gdscript
# ...

# The first player
var player_one : PlayerSelectChar = PlayerSelectChar.new()
# The second player
var player_two : PlayerSelectChar = PlayerSelectChar.new()

# ...

# Put this somewhere further in your script to initialize the two nodes
player_one.player_num = 0
player_two.player_num = 1

# ...

# Nodes do not need to be instantiated, only PackedScenes do that
# These lines assume that the current script is the parent that both nodes should belong to
# If it's a different script, then change the lines accordingly
add_child(player_one)
add_child(player_two)

# ...
```

Otherwise, if you want to do it in the editor, connect the scripts to the respective nodes that you wish to use as long as they're a Node2D or a derivative.

As for the latter, that is a little bit of a separate question. I'd assume that the engine should have the functionality for proper 2-player already implemented, but that is something that I haven't delved into yet.

**Reply by Raptorq1** [reply]
Oh it still doesn't work

**Reply by r6915ee** [reply]
Does the engine tell you anything about it? What does the issue look like?

**Reply by Raptorq1** [reply]
No it doesn't
![Screenshot_20241116-220924](https://github.com/user-attachments/assets/0ea3d49e-cd87-4270-8e28-7689d4921bdc)
![Screenshot_20241116-220951](https://github.com/user-attachments/assets/25ab877f-929b-40a6-b6da-4b9d8f8e2443)

**Reply by r6915ee** [reply]
If in the first image you're talking about the character icons, then a usual texture change should do the trick. And what about the second image?

**Reply by Raptorq1** [reply]
No its still not assigning the selected character to each player

**Reply by Caverns4** [reply]
In your menu, go to the point in the scene where the players have confirmed their choices and the level should start.
```
Global.PlayerChar1 = (Yourvar1)+1
Global.PlayerChar2 =  (Yourvar2)+1
```
This is assuming your first selectable player in the menus are entry 0. Replace the two "Yourvar"s with whatever variable you're using to store the selections.

**Reply by Raptorq1** [reply]
@Caverns4 I have added the code to the level start code but now it doesn't change scenes

extends Node2D

@export var nextZone =load("res://Data/Stages/TZ/Scene1.tscn")
@onready var player1 =$CanvasLayer/Player1
@onready var player2 =$CanvasLayer/Player2
var player1selected = player1.characterID
var player2selected = player2.characterID





func _process(delta: float) -> void:
	if player1.selected == true and player2.selected ==true:
		if Input.is_action_just_pressed("gm_pause"):
			Global.TwoPlayer = true
			Global.PlayerChar1 = player1selected+1
			Global.PlayerChar2 =  player2selected+1
			Global.main.change_scene_to_file(nextZone)

**Comment by Caverns4** [comment]
There is zero reason that should happen. Does the game crash or does it simply do nothing?

**Reply by Raptorq1** [reply]
It does nothing oh and when I test it in the editor it gives me this error Invalid access to property or key 'characterID' on a base object of type 'Nil'. in the debugger

**Reply by r6915ee** [reply]
I'm guessing for the second part that you referenced the nodes incorrectly in some way, shape, or form.

**Reply by Raptorq1** [reply]
> There is zero reason that should happen. Does the game crash or does it simply do nothing?

@Caverns4 I have a question how did you make the HUD show for both players oh and I fixed the issue

**Reply by Caverns4** [reply]
To get the HUD showing for both players, put simply, I didn't. The HUD in Sonic 2:Retold for Two Player Mode has new child node tree that's invisible by default and a script addition that activates it in two-player mode.

**Reply by Raptorq1** [reply]
@Caverns4 Oh and how did you get the 2 player HUD to work

**Reply by Raptorq1** [reply]
@Caverns4 how would I make it so that when the time over timer is at 10:00 it plays a sound

**Reply by Raptorq1** [reply]
@Caverns4 how would I adapt the script that you wrote for sonic 2 retold for my own game

Mania has 3 different screens it goes though before starting a two player level

character select 
![Screenshot_20250210-232230](https://github.com/user-attachments/assets/d5d3f5df-9932-4e08-870a-c5f53ae14bb2)

competition options
![Screenshot_20250210-232244](https://github.com/user-attachments/assets/5847faea-6ee8-4e10-a872-1e663c0a8f46)

And  finally zone select 
![Screenshot_20250210-232255](https://github.com/user-attachments/assets/d192dbeb-c030-4128-80fc-2772bf94a1e7)

**Comment by Caverns4** [comment]
Okay, so 1: I got the Two-Play HUD working by essentially making an entirely different HUD layer and making different thing visible/invisible on the ready function.
2: I don't really know why you'd want to do that since 10:00 is when the player dies. It'd be more practical to play the warning at 9:00. Either way, in the HUD script, you'll wand to preload the sound effect at the start of the script, and make a boolean to check if the warning sound has been triggered so it won't overlap, then, in _process, check if  (boolean is false AND the timer is >= 9*60). If true, Global.play_sound(warning) and bool = true
3: There are about a million different ways to do that. All you really need to do is make a separate screen for each screen, or use a Camera 2D to change what the player sees at any one time.

**Reply by Raptorq1** [reply]
Okay I was talking about the death timer

**Comment by DimensionWarped** [comment]
Raptor, your questions are inane. They are *always* inane. And frankly, I'm sick of seeing them. How do you make a character select screen? This is basic menu making. It's the simplest of the simple. You create an interface, the interface controls a variable. You create a similar interface when a second player needs to be selected, it controls another similar variable. Hell I think they are both elements of the same array if you are going by how the second player is controlled in the base engine.

I've told you a dozen times at this point that the discussion forum is only for thought out inquiries and "How do I do [x thing that can be done in literally infinite ways]" is not a thought out inquiry. If you can't figure out problems like this on your own, you need to stop screwing around with complicated engines someone else wrote and go learn how to use Godot. From the ground up. As a baseline amateur. The next time I see a thread like this, I'm closing it immediately.

**Reply by Raptorq1** [reply]
Sorry @DimensionWarped
